### PR TITLE
Add codegen to generate typed queries and mutations

### PR DESCRIPTION
**Pull request by mohit038-zz**
We can use graphql codegen to generate typed queries and mutations for the frontend.
So here is how it works.

We define all of our queries and mutations in SDL.

For example 
```
mutation Login($email: EmailAddress!, $password: String!) {
  login(email: $email, password: $password) {
    ... on UserReturned {
      user {
        id
      }
    }
    ... on InputValidationProblem {
      problems {
        path
        message
      }
    }
  }
}
 ```

Then run yarn ```yarn generate-api``` to generate typed hooks for Vue.
After that, we can import generated hooks in our components and use them like this.
```
const {
      mutate: loginUser,
      loading,
      error: graphqlError,
      onDone,
    } = useLoginMutation(() => ({
      variables: {
        email: email.value,
        password: password.value,
      },
    }))
```

It looks much cleaner than defining queries and mutations in the gql tag and using the UseQuery/mutation hook.

https://www.graphql-code-generator.com/docs/getting-started

### CODE REVIEW

**Comment by tobiasdiez** [comment]
Thanks a lot for your PR! 

Which advantages do you see in this solution using external graphql files over the graphql tag? What I liked about the current structure is that related things are coupled together. So if you change the view and, say, want to show an additional field, then you only need to change the vue file. Now you also have to look for the corresponding graphql file (and there is no easy way to jump from the vue to the correct graphql file).

File: codegen.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+overwrite: true
```
Comment (tobiasdiez) [inline codegen.yml:1]: There is already a similar code-gen config in `graphql.config.json`. Please merge these additions with that file (by modifying the content under `apollo`)

File: components/DocumentEditor.vue [hunk 2]
```diff
@@ -100,53 +100,9 @@
...
-    }
-    ... on Thesis {
-      institution {
-        id
-        name
-      }
-    }
-  }
-`)
+import { useDocumentDetailsQuery } from '~/types/graphql'
```
Comment (tobiasdiez) [inline components/DocumentEditor.vue:105]: Since these are not pure types, I would prefer if the generated code is placed under `~/composables` (or if that leads to conflicts under `~/apollo`)

File: components/DocumentEditor.vue [hunk 3]
```diff
@@ -216,7 +163,7 @@ export default defineComponent({
       published: computed({
         get: () =>
           document.value && 'published' in document.value
-            ? document.value.published
+            ? (document.value.published as Date)
```
Comment (tobiasdiez) [inline components/DocumentEditor.vue:166]: This change should no longer be necessary after merging the code gen config

File: components/NavBar.vue [hunk 4]
```diff
@@ -107,37 +107,27 @@
...
-      void resolveClient().clearStore()
+    const { mutate: logout, loading, error, onDone } = useLogoutMutation()
 
-      void router.push({ name: 'index' })
+    const router = useRouter()
+    onDone((result) => {
+      if (result.data?.logout) {
+        // Reset graphql cache
+        void resolveClient().clearStore()
+        cacheCurrentUser(null)
```
Comment (tobiasdiez) [inline components/NavBar.vue:128]: Is this really the "correct" way to handle the cache? Normally, apollo has these update methods that work slightly different then the onDone callers.

File: graphql-api/fragments/DocumentDetails.graphql [hunk 5]
```diff
@@ -0,0 +1,41 @@
+fragment DocumentDetails on Document {
```
Comment (tobiasdiez) [inline graphql-api/fragments/DocumentDetails.graphql:1]: Please put these files right next to the vue files that use them, and maybe rename them, so that its relatively easy to jump to the correct graphql file.

File: package.json [hunk 6]
```diff
@@ -37,7 +37,8 @@
     "test:integration": "cross-env NODE_OPTIONS=--experimental-vm-modules jest test.ts --runInBand",
     "test:unit": "cross-env NODE_OPTIONS=--experimental-vm-modules jest spec.ts",
     "typecheck": "nuxi typecheck",
-    "validate": "yarn graphql:validate"
+    "validate": "yarn graphql:validate",
+    "generate-api": "graphql-codegen --config codegen.yml"
```
Comment (tobiasdiez) [inline package.json:41]: Also shouldn't be necessary after merging the configs (done via the graphql:generate command defined above)

File: package.json [hunk 7]
```diff
@@ -91,13 +92,14 @@
   },
   "devDependencies": {
     "@babel/core": "^7.17.5",
-    "@graphql-codegen/cli": "^2.6.2",
+    "@graphql-codegen/cli": "2.6.2",
```
Comment (tobiasdiez) [inline package.json:95]: Please revert these changes to the version constraints.

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
While I still like the colocation of graphql queries in the vue files, the dev experience of this solution is still not great: no intellisense, problems with linters etc. So lets go with your solution for now, and reconsider this when the ecosystem is more mature.

I have a few points that need improvements. Could you please have a look? Thanks!

File: components/NavBar.vue (hunk 4 above)
Comment (mohit038) [reply components/NavBar.vue:141]: Actually, Updating the cache here is kind of unnecessary. This would be helpful if the data fetched by me query is the same as the data we get from login mutation(i.e., if we only query for email or id in user field with me query, then we are avoiding a request to the server). But right after the login, we query for user document with me query, which will result in a cache miss, and apollo client will send another request to the server. Let me know if I'm missing something, but I don't think we are avoiding any requests here.

File: graphql-api/fragments/DocumentDetails.graphql (hunk 5 above)
Comment (mohit038) [reply graphql-api/fragments/DocumentDetails.graphql:1]: I'm not sure about this, but wouldn't it conflict with the react-router if we put graphql files in pages?

File: components/NavBar.vue (hunk 4 above)
Comment (tobiasdiez) [reply components/NavBar.vue:141]: I agree with your analysis for what happens at the login. However, here we logout the user and need to tell graphql that this means `user = null` (otherwise the apollo client accesses the cache and thinks the user is still logged in).

Comment (tobiasdiez) [reply components/NavBar.vue:141]: And even for the login, there are a few requests that profit from caching such as this one here:
https://github.com/JabRef/JabRefOnline/blob/b279d4a1e18e93d5d0670fb550224e7b00948f2c/plugins/authenticated.ts#L19-L27
(which is run on every page load to check if the user is logged in).

**Comment by tobiasdiez** [comment]
Linters and intellisense (https://github.com/graphql/graphiql/pull/2827) are now working in vue files as well, so I don't see little value in declaring the graphql queries outside of the components. If you need to reuse a query in multiple places, you can easily extract it into a composable (=hook) or create a new reusable component.

Nonetheless, thanks a lot @mohit038 for your work on this!
